### PR TITLE
8205592: BigDecimal.doubleValue() is depressingly slow

### PR DESCRIPTION
**Pull request by rgiulietti**
A reimplementation of `BigDecimal.[double|float]Value()` to enhance performance, avoiding an intermediate string and its subsequent parsing on the slow path.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8205592](https://bugs.openjdk.org/browse/JDK-8205592): BigDecimal.doubleValue() is depressingly slow

### Reviewers
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to [75fde901](https://git.openjdk.org/jdk/pull/9410/files/75fde90104de71365d58dac46c776946e4b65e68)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/9410/head:pull/9410` \
`$ git checkout pull/9410`

Update a local copy of the PR: \
`$ git checkout pull/9410` \
`$ git pull https://git.openjdk.org/jdk.git pull/9410/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9410`

View PR using the GUI difftool: \
`$ git pr show -t 9410`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9410.diff">https://git.openjdk.org/jdk/pull/9410.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/9410#issuecomment-1177797208)

### CODE REVIEW

**Comment by TheShermanTanker** [comment]
Impressive performance improvements, I also like the PR title :P

**Comment by rgiulietti** [comment]
@TheShermanTanker
The title is the one from the JBS issue.
As for the performance improvements, note that microbenchmarks are what they are. The overall effect of these improvements might be barely noticeable in some programs using BigDecimal and might be positively impact others.

**Comment by rgiulietti** [comment]
keep open

**Comment by rgiulietti** [comment]
keep open, waiting for review

**Comment by rgiulietti** [comment]
waiting for review

**Comment by rgiulietti** [comment]
waiting for review

**Comment by rgiulietti** [comment]
This PR is waiting for a review

**Comment by plokhotnyuk** [comment]
@rgiulietti Thanks for keeping making JDK faster!

I have a couple of review comments:

1) For the line https://github.com/openjdk/jdk/pull/9410/files#diff-94d400b99466045dd76001c37eada6b24d086d8d115b49c439752bbceb233772L3746

It seems that removing this check for 22-bit number introduces an extra rounding error.

Please consider just increasing the constant up to 1L << 32 and using DOUBLE_10_POW instead of FLOAT_10_POW.

Bellow are REPL expressions that illustrate my thoughts (sorry for using Scala):
```
$ scala
Welcome to Scala 3.2.0 (17.0.5, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                                                                                                   
scala> new java.math.BigDecimal("167772170E-1").floatValue
val res0: Float = 1.6777216E7
                                                                                                                                                                                                                   
scala> 167772170E-1f
val res1: Float = 1.6777216E7
                                                                                                                                                                                                                   
scala> 167772170 / 10.0f
val res2: Float = 1.6777218E7

scala> (167772170 / 10.0).toFloat
val res3: Float = 1.6777216E7
```
2) For the line: https://github.com/openjdk/jdk/pull/9410/files#diff-94d400b99466045dd76001c37eada6b24d086d8d115b49c439752bbceb233772L3791

It seems that removing this check for 52-bit number introduces an extra rounding error:
```
$ scala
Welcome to Scala 3.2.0 (17.0.6, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.

scala> new java.math.BigDecimal("90071992547409930E-1").doubleValue
val res0: Double = 9.007199254740992E15
                                                                                                                                                                                                                   
scala> 90071992547409930E-1
val res1: Double = 9.007199254740992E15
                                                                                                                                                                                                                   
scala> 90071992547409930L / 10.0
val res2: Double = 9.007199254740994E15
```
3) For the line: https://github.com/openjdk/jdk/pull/9410/files#diff-94d400b99466045dd76001c37eada6b24d086d8d115b49c439752bbceb233772L3800
Here before falling back to the costly fallback with big number operations we can [get a number of digits in intCompact](https://github.com/plokhotnyuk/jsoniter-scala/blob/6f702ce5cae05df91b5aa6e4bd61acdf43bf18f6/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L1930) and use [a trick with two multiplications](https://github.com/plokhotnyuk/jsoniter-scala/blob/6f702ce5cae05df91b5aa6e4bd61acdf43bf18f6/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1459)

**Comment by plokhotnyuk** [comment]
Sorry, I failed to recognize those checks two times :)

How about adding [this moderate path](https://github.com/plokhotnyuk/jsoniter-scala/blob/6f702ce5cae05df91b5aa6e4bd61acdf43bf18f6/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1467-L1498)?

Also, I think it worth do be reused for regular parsing of double and float values from string.

**Comment by rgiulietti** [comment]
@plokhotnyuk The main goal of this PR is to avoid generating a string and parse it, as it happens in the current implementation. The fact that it results in being faster is only a welcome byproduct.

The proposed patch for `doubleValue()` is only about 40 lines of code, not counting `}`-only lines and the _extensive comments_ explaining the details for the benefit of both reviewers and maintainers. Shorter, documented code has a higher chance to be correct and understood. It also contributes to simpler and quicker reviewing.

Adding the "moderate path" to this patch would increase the code size considerably. Moreover, I would have to invest time to understand the dense, uncommented code and convince myself and the reviewers that it is correct. I would also have to setup benchmarks to measure the overall benefits of adding it to the proposed patch. And add specific tests to cover the path.

Before that, I would prefer for this patch to be first reviewed as it is (with possible corrections). I hope to have time to invest into your proposals _once_ this PR is integrated into mainline. Thanks for your patience.

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (jddarcy) [inline src/java.base/share/classes/java/math/BigDecimal.java:308]: It doesn't matter in terms of the code, but shouldn't this be something like:

L - (ulp(L)) < l < L  ulp(L)

In other words, without further checking, it isn't clear that L is the lower-bound of the two double value bracketing l.

(If the ulp function being discussed were the real-valued version than L +/- ulp(l) would also be a reasonable formulation.)

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (jddarcy) [inline src/java.base/share/classes/java/math/BigDecimal.java:3777]: I'd prefer a name other than "d" be used for the BigInteger significand's magnitude.

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (jddarcy) [inline src/java.base/share/classes/java/math/BigDecimal.java:3749]: If the total size of code were a concern, I wouldn't be adverse to floatValue() being implemented as

return (float)doubleValue();

Given the 2p+2 property between the float and double formats, BigDecimal -> double -> float and BigDecimal -> float round the same way.

**Review by jddarcy — APPROVED** [review]

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigDecimal.java:308]: While it isn't obvious from the definition, high precision computations reveal that the inequalities `L < l < L + ulp(L) / 2` do indeed hold.
But as noted, this is not really relevant for the rest of the analysis.

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigDecimal.java:3749]: No, that would not be correct. It would be subject to [double rounding](https://en.wikipedia.org/wiki/Rounding#Double_rounding), against the spec.

For example, `BigDecimal` 1.000000059604644775390626 should round to `float` `1.0000001f`.
When going to the closest `double` and then to the closest `float`, however, it first rounds to `double` `1.0000000596046448` and the to `float` `1.0f`.

File: src/java.base/share/classes/java/math/BigDecimal.java
Comment (jddarcy) [reply src/java.base/share/classes/java/math/BigDecimal.java:3777]: > I have no objections as long as it is a 1 letter name that does not conflict with others. Have you a preferred one?

I usually use "bd" or "bi" for such BigDecima/BigInteger variables myself and use "d" for double; "v" or "w" have integer-ish feel to me so perhaps one of those?

**Pull request closed without merging**
